numpy/core/src/common/npstd.hpp

```diff
@@ -34,23 +36,439 @@ using std::complex;
+    {
+        return bits_ == r.bits_ && IsFinite() && r.IsFinite();
+    }
+    constexpr bool operator!=(const AbstractTime &r) const
+    {
+        return bits_ != r.bits_ || IsNaT() || r.IsNaT();
+    }
+    constexpr bool operator<(const AbstractTime &r) const
+    {
+        return bits_ < r.bits_ && IsFinite() && r.IsFinite();
+    }
+    constexpr bool operator>(const AbstractTime &r) const
+    {
+        return r < *this;
+    }
+    constexpr bool operator<=(const AbstractTime &r) const
+    {
+        return bits_ <= r.bits_ && IsFinite() && r.IsFinite();
+    }
+    constexpr bool operator>=(const AbstractTime &r) const
+    {
+        return r <= *this;
+    }
+    /// @}
+
+    /// @name Arithmetic operators
+    /// @{
+    constexpr AbstractTime operator+(const AbstractTime &r) const
+    {
+        if (IsNaT() || r.IsNaT()) {
+            return NaT();
+        }
+        return AbstractTime(bits_ + r.bits_);
+    }
+    constexpr AbstractTime &operator+=(const AbstractTime &r)
+    {
+        *this = (*this) + r;
+        return *this;
+    }
+    constexpr AbstractTime operator-(const AbstractTime &r) const
+    {
+        if (IsNaT() || r.IsNaT()) {
+            return NaT();
+        }
+        return AbstractTime(bits_ - r.bits_);
+    }
+    constexpr AbstractTime &operator-=(const AbstractTime &r)
+    {
+        *this = (*this) - r;
+        return *this;
+    }
+    constexpr AbstractTime operator*(const AbstractTime &r) const
+    {
+        if (IsNaT() || r.IsNaT()) {
+            return NaT();
+        }
+        return AbstractTime(bits_ * r.bits_);
+    }
+    constexpr AbstractTime &operator*=(const AbstractTime &r)
+    {
+        *this = (*this) * r;
+        return *this;
+    }
+    constexpr AbstractTime operator/(const AbstractTime &r) const
+    {
+        if (IsNaT() || r.IsNaT()) {
+            return NaT();
+        }
+        return AbstractTime(bits_ / r.bits_);
+    }
+    constexpr AbstractTime &operator/=(const AbstractTime &r)
+    {
+        *this = (*this) / r;
+        return *this;
+    }
+    /// @}
+
+    /// @name Properties
+    /// @{
+    constexpr bool IsNaT() const
+    {
+        return bits_ == static_cast<int64_t>(NaT());
+    }
+    constexpr bool IsFinite() const
+    {
+        return bits_ != static_cast<int64_t>(NaT());
+    }
+    /// @}
+
+  private:
+    int64_t bits_;
+};
+
+/// Holds dates or datetimes with a precision based on selectable date or time units.
+class DateTime : public AbstractTime
+{
+  public:
+    using AbstractTime::AbstractTime;
+};
+/// Holds lengths of times in integers of selectable date or time units.
+class TimeDelta : public AbstractTime
+{
+  public:
+    using AbstractTime::AbstractTime;
+};
+
+/** Hold the reference of arbitrary Python object.
  *
- * This type will be set to `void` when `npy_longdouble` is not defined
- * as `long double`.
+ * Note: this class doesn't handel reference counting
+ * via RAII or raise any kind of exceptions.
```

Comment:
Why not? It seems a bit of a trap at the moment, because `Object(1);` leaks a reference, but `Object(some_obj)` doesn't.

Reply:
Supporting RAII should clean up the code, I'm going to support it. My reason at the moment is that we disable C++ exceptions by default which means there's no support for STD runtime including memory operators e.g. new/delete.